fs/walk.ts

```diff
@@ -11,6 +11,16 @@ import {
   WalkEntry,
 } from "./_util.ts";
 
+export class WalkError extends Error {
+  path: string;
+  constructor(err: unknown, path: string) {
+    super(`${err instanceof Error ? err.message : err} for path "${path}"`);
+    this.name = "WalkError";
+    this.path = path;
+    this.cause = err;
+  }
```

Comment:
```suggestion
  override cause: unknown;
  override name = "WalkError";
  path: string;

  constructor(cause: unknown, path: string) {
    super(
      `${cause instanceof Error ? cause.message : cause} for path "${path}"`,
    );
    this.path = path;
    this.cause = cause;
  }
```

See usage info for `override` in the [TS handbook](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-3.html#override-and-the---noimplicitoverride-flag), and other usage examples in this repo — for example, in `AssertionError` at [`std/testing/asserts.ts#L20`](https://github.com/denoland/deno_std/blob/0.178.0/testing/asserts.ts#L20).

This also improves typing for this error subclass by removing the [optional modifier](https://www.typescriptlang.org/docs/handbook/2/objects.html#optional-properties) from the `cause` property — because it is guaranteed to be set to some (`unknown`) value.

Reply:
Thanks for the review, I think this makes sense. I don't think this is mergable because I don't know if this is testable. If you could come up with a test case to test this code then I would love to move this PR forward.

Reply:
@lino-levan I don't have write permission for this PR to update the test file (I'm not an org member, nor am I a collaborator on your branch) — so I can't contribute a commit to revise it. FYI: There's a types assertion module in this repo at [`std/testing/types.ts`](https://github.com/denoland/deno_std/blob/main/testing/types.ts).

Reply:
Ah sorry, I wasn't being super clear. This PR in general (outside of this change) is not landable because I can't write tests for it. As far as I can tell, this error class only gets used when some race conditions occur (and those can't be reproc'd). I would change this PR to a draft but I can't seem to figure out how to do that.

Reply:
Understood — but what's holding you back? Your [initial idea](https://github.com/denoland/deno_std/pull/3054#issuecomment-1405840646) seemed like a contrived, but perfectly fine approach for generating an exception in a test.

Reply:
Fair enough. I'll look into it this weekend. This change looks good, I'll commit it. Thanks again for the review!

Reply:
> I tried my hand at writing a test for this but can't think of a way for this to error out consistently (without relying on something like removing a folder after we already started walking it). Do you have any ideas?

@lino-levan I just tested this locally and it reliably throws every run:

`test.ts`:
```ts
// deno test --allow-read=exampleRootTestDir --allow-write=exampleRootTestDir

import { assertRejects } from "https://deno.land/std@0.178.0/testing/asserts.ts";
import { walk } from "https://deno.land/std@0.178.0/fs/walk.ts";

Deno.test("throws", async () => {
  const dir = "exampleRootTestDir";

  await assertRejects(async () => {
    await Deno.mkdir(dir);
    for await (const _walkEntry of walk(dir)) await Deno.remove(dir);
  });
});

```

```
 % deno --version
deno 1.31.1 (release, aarch64-apple-darwin)
v8 11.0.226.13
typescript 4.9.4

% deno test --allow-read=exampleRootTestDir --allow-write=exampleRootTestDir
running 1 test from ./test.ts
throws ... ok (6ms)

ok | 1 passed | 0 failed (20ms)
```

Check out the [`testWalk`](https://github.com/denoland/deno_std/blob/0.178.0/fs/walk_test.ts#L11) function in the test file.